### PR TITLE
feat: [CDS-70144]: Allow branch selection while using templates in template/pipeline studio

### PR DESCRIPTION
**Pull request by sonaligoyal-harness**
### Summary

For remote pipeline/template, if user wants to use template inside it, it could be of branch of which pipeline/template has been created.
If user selects template of different branch then by default from BE, by default branch was chosen and template of different repo was linked to parent template/ pipeline
 But with this enhancement, we are allowing user to select any branch from any repo to link in their template/ pipeline.
 
 **Default behaviour( Send parentEntityDetails in api call and No need to show gitBranch)**

Case 1: Remote pipeline(Parent) is in repo A and default branch.
- If user selects same repo A template to link inside it, then by default, default branch will be selected.
- If user selects different repo B template to link inside it, then by default, default branch of repo B will be selected. 

Case 2: Remote pipeline(Parent) is in repo A and feature branch.
- If user selects same repo A template to link inside it, then by default, feature branch will be selected.
- If user selects different repo B Template to link inside it, then by default, default branch of repo B will be selected.

**Static Behaviour(Send only selectedBranch in api call and show gitBranch in yaml)**

 When user selects branch different from the branch mentioned in above cases, it will lead to static behaviour and in these cases , we need special handling for all api calls

#### Screenshots

Same repo default branch (default behaviour)

https://github.com/harness/harness-core-ui/assets/98508399/caa0f018-4ca5-478f-8c30-ed74a4e0c3ed

Same repo different branch( Static behaviour)

https://github.com/harness/harness-core-ui/assets/98508399/7f470a29-5e08-4eb9-81e2-5ac2b5d3d0ee

Different repo default branch ( Default behaviour)

https://github.com/harness/harness-core-ui/assets/98508399/806f7a68-3ff5-4269-a44b-adefc90210c8

Different repo different branch (static behaviour)

https://github.com/harness/harness-core-ui/assets/98508399/1be7addb-78f8-4cb5-9f72-c875d1e3cd5a

**Error handling :** 

https://github.com/harness/harness-core-ui/assets/63278928/badd369b-5eac-48e1-8b2b-ef8d9b1f1bd9

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

File: src/modules/10-common/components/GitPopoverV2/GitPopoverV2.tsx [hunk 1]
```diff
@@ -131,15 +135,21 @@ export const GitPopoverV2 = ({
     }
   }, [forceFetch, branchChangeDisabled, refetch])
 
+  useEffect(() => {
+    //here, default branch is set which is getting consumed in template details for setting it in selected template gitdetails object
+    if (response?.data?.defaultBranch?.name) setDefaultBranch?.(response?.data?.defaultBranch?.name)
```
Comment (pankajsingh-harness) [inline src/modules/10-common/components/GitPopoverV2/GitPopoverV2.tsx:140]: NIT: This optional chaining can be removed as the value is already checked. 
```suggestion
    if (response?.data?.defaultBranch?.name) setDefaultBranch?.(response.data.defaultBranch.name)
```

File: src/modules/10-common/components/GitPopoverV2/GitPopoverV2.tsx [hunk 2]
```diff
@@ -131,15 +135,21 @@ export const GitPopoverV2 = ({
...
     if (response?.status === 'SUCCESS' && !isEmpty(response?.data)) {
       return getBranchSelectOptions(response?.data?.branches)
     }
     return []
   }, [response?.data, response?.status])
 
-  const selectedBranch = createSelectOption(defaultTo(branch, response?.data?.defaultBranch?.name))
-
+  const selectedBranch = createSelectOption(
+    defaultTo(selectedBranchInPopover, defaultTo(branch, response?.data?.defaultBranch?.name))
```
Comment (pankajsingh-harness) [inline src/modules/10-common/components/GitPopoverV2/GitPopoverV2.tsx:151]: I think second `defaultTo` is not needed as `response?.data?.defaultBranch?.name` is used to set the state value for selectedBranchInPopover

File: src/modules/10-common/utils/gitSyncUtils.ts [hunk 3]
```diff
@@ -55,9 +57,10 @@ export const getGitQueryParamsWithParentScope = ({
     branch: branchParam,
 
     // Git experience uses storeMetadata
-    parentEntityConnectorRef: storeMetadata?.connectorRef,
-    parentEntityRepoName: storeMetadata?.repoName,
-    ...(!isEmpty(storeMetadata?.connectorRef) ? parentEntityIds : {}),
+    ...(sendParentEntityDetails
+      ? { parentEntityConnectorRef: storeMetadata?.connectorRef, parentEntityRepoName: storeMetadata?.repoName }
+      : {}),
+    ...(!isEmpty(storeMetadata?.connectorRef) && sendParentEntityDetails ? parentEntityIds : {}),
```
Comment (pankajsingh-harness) [inline src/modules/10-common/utils/gitSyncUtils.ts:63]: NIT: This condition can be moved to upper condition

File: src/modules/70-pipeline/components/PipelineStudio/PipelineContext/PipelineContext.tsx [hunk 4]
```diff
@@ -1236,21 +1237,29 @@ export function PipelineProvider({
       return
     }
 
+    const getBranchForSelectedStage = () => {
+      return getStageFromPipeline(state.selectionState?.selectedStageId as string)?.stage?.stage?.template?.gitBranch
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineContext/PipelineContext.tsx:1241]: `getStageFromPipeline(state.selectionState?.selectedStageId as string)?.stage?.stage?.template?.gitBranch` can be stored in a const and then used at both place.

File: src/modules/70-pipeline/utils/templateUtils.tsx [hunk 5]
```diff
@@ -98,6 +135,15 @@ export const createTemplate = <T extends PipelineInfoConfig | StageElementConfig
       if (template.versionLabel) {
         set(draft, 'template.versionLabel', template.versionLabel)
       }
+      if (template.gitDetails?.branch && template.gitDetails?.repoName) {
+        if (
```
Comment (pankajsingh-harness) [inline src/modules/70-pipeline/utils/templateUtils.tsx:139]: This if conditions can be move to the upper if with &&. 
```suggestion
        if ((template.gitDetails?.branch && template.gitDetails?.repoName) && ((template.gitDetails.repoName === repoName && template.gitDetails.branch !== branch) ||
          (template.gitDetails.repoName !== repoName &&
            template.gitDetails.branch !== template.gitDetails?.defaultBranch)))
```

File: src/modules/72-templates-library/components/TemplatePipelineContext/TemplatePipelineContext.tsx [hunk 6]
```diff
@@ -146,7 +146,15 @@ export function TemplatePipelineProvider({
         storeMetadata
       })
     )
+
+    const getBranchForSelectedStage = () => {
+      return getStageFromPipeline(state.selectionState?.selectedStageId as string)?.stage?.stage?.template?.gitBranch
```
Comment (pankajsingh-harness) [inline src/modules/72-templates-library/components/TemplatePipelineContext/TemplatePipelineContext.tsx:151]: Same as above

Comment (pankajsingh-harness) [reply src/modules/72-templates-library/components/TemplatePipelineContext/TemplatePipelineContext.tsx:151]: This logic can be moved to an util function and used at both the places.

**Review by Eric761 — DISMISSED** [review]

File: src/modules/70-pipeline/components/PipelineStudio/PipelineContext/PipelineContext.tsx [hunk 7]
```diff
@@ -398,6 +399,7 @@ const getTemplateType = (
   loadFromCache?: boolean
 ): ReturnType<typeof getTemplateTypesByRef> => {
   const templateRefs = uniq(findAllByKey('templateRef', pipeline))
+  const gitBranches = extractGitBranchUsingTemplateRef(pipeline, '')
```
Comment (sunnykesh-harness) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineContext/PipelineContext.tsx:402]: We can rename it to templateGitBranches.

**Review by sunnykesh-harness — COMMENTED** [review]
LGTM

File: src/modules/10-common/utils/gitSyncUtils.ts (hunk 3 above)
Comment (sonaligoyal-harness) [reply src/modules/10-common/utils/gitSyncUtils.ts:63]: No, this has different use case and we can't combine that

**Review by sunnykesh-harness — APPROVED** [review]

**Review by pankajsingh-harness — APPROVED** [review]

**Review by bhavya-sinha — APPROVED** [review]

**Review by vardanbansal-harness — APPROVED** [review]

**Pull request merged**